File: src/lib/db/playlist.ts
// ********************************************
// ************** Playlist stuff **************
// ********************************************

export type OldImageDB = {
    url: string;
    width: number;
    height: number;
};

export type ArtistDB = {
    name: string;
    id: string;
};

import type { SongDB, SongDBFull } from "./song";

export type PlaylistDB<
    Keys extends keyof PlaylistDBFull = keyof PlaylistDBFull,
> = Pick<PlaylistDBFull, Keys>;

export interface RawPlaylistDB {
    id: string;
    images: string | undefined;
    image: string;
    name: string;
    description: string;
    owner: string;
    followers: number;
    songs: string;
    updatedAt: string | undefined;
    createdAt: string | undefined;
}

export interface PlaylistDBFull {
    id: string;
    images: OldImageDB[] | undefined;
    image: string;
    name: string;
    description: string;
    owner: string;
    followers: number;
    songs: PlaylistDBSong[];
    type: "playlist";
    updatedAt: string | undefined;
    createdAt: string | undefined;
}

export interface PlaylistDBSong {
    id: string;
    added_at?: string;
    addedInRockit?: boolean | undefined;
}

export type PlaylistDBSongWithAddedAt<
    Keys extends keyof SongDBFull = keyof SongDBFull,
> = SongDB<Keys> & {
    added_at?: string;
};

export function parsePlaylist(
    playlist: RawPlaylistDB | undefined
): PlaylistDB | undefined {
    if (!playlist) return undefined;

    const out = {
        id: playlist.id,
        images: JSON.parse(playlist.images || "[]"),
        image: playlist.image,
        name: playlist.name,
        description: playlist.description,
        owner: playlist.owner,
        followers: playlist.followers,
        songs: JSON.parse(playlist.songs || "[]"),
        updatedAt: playlist.updatedAt,
        createdAt: playlist.createdAt,
        type: "playlist" as const,
    };

    Object.entries(out).forEach((entry) => {
        if (typeof entry[1] == "undefined") {
            // @ts-expect-error delete undefined properties
            delete out[entry[0]];
        }
    });

    return out;
}

export const playlistQuery = `CREATE TABLE IF NOT EXISTS playlist (
    id TEXT NOT NULL PRIMARY KEY UNIQUE,
    images TEXT,
    image TEXT NOT NULL DEFAULT "",
    name TEXT NOT NULL,
    description TEXT NOT NULL,
    owner TEXT NOT NULL,
    followers INTEGER NOT NULL,
    songs TEXT NOT NULL,
    updatedAt TEXT,
    createdAt TEXT
)`;

File: backend/db/playlist.py
from typing import TypedDict, List, Optional
import json
from dataclasses import dataclass

from backend.db.commonTypes import OldImageDB


class RawPlaylistDB(TypedDict):
    id: str
    images: Optional[str]
    image: str
    name: str
    description: str
    owner: str
    followers: int
    songs: str
    updatedAt: Optional[str]
    createdAt: Optional[str]


@dataclass
class PlaylistDBSong:
    id: str
    added_at: Optional[str]
    addedInRockit: Optional[bool] = None


@dataclass
class PlaylistDBFull:
    id: str
    images: Optional[List[OldImageDB]]
    image: str
    name: str
    description: str
    owner: str
    followers: int
    songs: List[PlaylistDBSong]
    updatedAt: Optional[str]
    createdAt: Optional[str]


def parse_playlist(raw_playlist: Optional[RawPlaylistDB]) -> Optional[PlaylistDBFull]:
    if not raw_playlist:
        return None

    raw_playlist_images = raw_playlist.get("images", "[]")

    return PlaylistDBFull(
        id=raw_playlist.get("id"),
        images=json.loads(
            s=raw_playlist_images) if raw_playlist_images else None,
        image=raw_playlist.get("image"),
        name=raw_playlist.get("name"),
        description=raw_playlist.get("description"),
        owner=raw_playlist.get("owner"),
        followers=raw_playlist.get("followers"),
        songs=[PlaylistDBSong(**args)
               for args in json.loads(raw_playlist.get("songs", "[]"))],
        updatedAt=raw_playlist.get("updatedAt"),
        createdAt=raw_playlist.get("createdAt"),
    )


playlist_query = """
CREATE TABLE IF NOT EXISTS playlist (
    id TEXT NOT NULL PRIMARY KEY UNIQUE,
    images TEXT,
    image TEXT NOT NULL DEFAULT "",
    name TEXT NOT NULL,
    description TEXT NOT NULL,
    owner TEXT NOT NULL,
    followers INTEGER NOT NULL,
    songs TEXT NOT NULL,
    updatedAt TEXT,
    createdAt TEXT
)
"""
